windows/src/desktop/kmshell/main/Keyman.System.UpdateStateMachine.pas

```diff
@@ -916,67 +980,156 @@ procedure WaitingRestartState.HandleInstallNow;
-  // for now we only want the exe although excute install can
-  // handle msi
+  found := False;
   for FileName in FileNames do
   begin
     fileExt := LowerCase(ExtractFileExt(FileName));
     if fileExt = '.exe' then
+    begin
+      found := True;
       break;
+    end;
   end;
 
-  if DoInstallKeyman(SavePath + ExtractFileName(FileName)) then
-  begin
-    KL.Log('TUpdateStateMachine.InstallingState.Enter: DoInstall OK');
-  end
+  // switch -au for auto update in silent mode.
+  // We will need to add the pop up that says install update now yes/no
+  // This will run the setup executable which will ask for  elevated permissions
+  if found then
+    FResult := TUtilExecute.Shell(0, SavePath + ExtractFileName(FileName),
+      '', '-au')
   else
+    FResult := False;
+
+  if not FResult then
   begin
-    // TODO: #10210 clean failed download
-    // TODO: #10210 Do we do a retry on install? probably not
-    KL.Log('TUpdateStateMachine.InstallingState.Enter: DoInstall fail');
+    bucStateContext.RemoveCachedFiles;
+    TKeymanSentryClient.Client.MessageEvent(Sentry.Client.SENTRY_LEVEL_ERROR,
+      'Executing kmshell process to install failed:"' +
+      IntToStr(Ord(FResult)) + '"');
     ChangeState(IdleState);
-  end
+  end;
+
+  Result := FResult;
+end;
+
+function InstallingState.DoInstallPackage(PackageFileName: String): Boolean;
+var
+  FPackage: IKeymanPackageFile2;
+begin
+  Result := True;
+  KL.Log('InstallingState.DoInstallPackage Entry' + PackageFileName);
+  FPackage := kmcom.Packages.GetPackageFromFile(PackageFileName)
+    as IKeymanPackageFile2;
+  FPackage.Install2(True);
+  // Force overwrites existing package and leaves most settings for it intact
+  FPackage := nil;
+
+  kmcom.Refresh;
+  kmcom.Apply;
+  KL.Log('InstallingState.DoInstallPackage about to delete');
+  System.SysUtils.DeleteFile(PackageFileName);
+
+end;
+
+function InstallingState.DoInstallPackages
+  (Params: TUpdateCheckResponse): Boolean;
+var
+  i: Integer;
+  SavePath: String;
+  PackageFullPath: String;
+begin
+  SavePath := IncludeTrailingPathDelimiter(TKeymanPaths.KeymanUpdateCachePath);
+  for i := 0 to High(Params.Packages) do
+  begin
+    PackageFullPath := SavePath + Params.Packages[i].FileName;
+    if not DoInstallPackage(PackageFullPath) then // I2742
+    begin
+      // Package did install log or error
+      KL.Log('Installing Package failed' + PackageFullPath);
+    end;
+  end;
+  Result := True;
+end;
+
+procedure InstallingState.Enter;
+var
+  ucr: TUpdateCheckResponse;
+  hasPackages, hasKeymanInstall: Boolean;
+begin
+
+  hasPackages := False;
+  hasKeymanInstall := False;
+  bucStateContext.SetRegistryState(usInstalling);
+
+  if (TUpdateCheckStorage.LoadUpdateCacheData(ucr)) then
+  begin
+    hasPackages := TUpdateCheckStorage.HasKeyboardPackages(ucr);
+    hasKeymanInstall := TUpdateCheckStorage.HasKeymanInstallFile(ucr);
+  end;
+  KL.Log('InstallingState.Enter before hasPackages');
+  if hasPackages then
```

Comment:
The order of first checking for `hasPackages` and then `hasKeymanInstall` looks illogical to me since I don't think you can install packages without having Keyman...

Reply:
This is for updating not for installing Keyman the first time the Keyman setup executable does that, even with this update state machine it will launch the setup executable. 
Keyman will be available if this code (kmshell) is exectuing.
It is logical to update Keyman first before packages I agree. Except long story.. we are trying to reduce the number of times the user has to be asked to elevate to admin or restart. 
Keyboards always need elevation, when we do that and execute kmshell as an elevated process we can then launch the Keyman installer and it will no longer ask for elevation.